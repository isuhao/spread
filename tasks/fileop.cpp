#include "fileop.hpp"

#include <boost/filesystem.hpp>

using namespace Tasks;
using namespace Jobify;
namespace bs = boost::filesystem;

void FileOpTask::addOp(const FileOp &op)
{
  ops.push_back(op);
}

void FileOpTask::addOp(int type, const std::string &source, const std::string &dest)
{
  FileOp op;
  op.type = type;
  op.source = source;
  op.dest = dest;
  addOp(op);
}

void FileOpTask::doJob()
{
  info->total = ops.size();
  info->current = 0;

  OpList::iterator it;
  for(it = ops.begin(); it != ops.end(); it++)
    {
      const FileOp &op = *it;

      // Deletes are easy to handle, so get them out of the way first.
      if(op.type == FILEOP_DELETE)
        {
          setBusy("Removing " + op.dest);
          bs::remove_all(op.dest);
          continue;
        }

      /* Fix this up later. I want more robust wrt directory copying
         and moving. In the copy case, it is pretty straight forward
         to index a directory (weeding out symbolic link loops), then
         copy each file individually, making sure to write directories
         along the way.
       */
      if(op.type == FILEOP_MOVE)
        {
          setBusy("Moving " + op.source + " => " + op.dest);
          if(bs::exists(op.dest) && !bs::is_directory(op.dest))
            bs::remove(op.dest);
          bs::rename(op.source, op.dest);
        }
      else if(op.type == FILEOP_COPY)
        {
          // Determine what to do first
          if(!bs::exists(op.source))
            setError("File not found - " + op.source);
          else if(bs::is_directory(op.source))
            setError("Directory copy not implemented yet");
          else
            {
              // Source exists and is a file
              if(bs::exists(op.dest))
                {
                  // Destination exists
                  if(bs::is_directory(op.dest))
                    setError("Cannot copy to directories yet");
                  else
                    // Overwrite existing file
                    bs::remove(op.dest);
                }
            }

          // Exit on errors
          if(!info->isBusy())
            return;

          // Single file copy only
          setBusy("Copying " + op.source + " => " + op.dest);
          bs::copy_file(op.source, op.dest);
        }
      else setError("Invalid operation");

      // Check for aborts or errors
      if(checkStatus())
        return;

      info->current++;
    }

  setDone();
}
