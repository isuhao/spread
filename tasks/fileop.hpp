#ifndef __TASKS_FILEOP_HPP_
#define __TASKS_FILEOP_HPP_

#include <job/job.hpp>
#include <list>

namespace Tasks
{
  /* File operations task. Can be used to run large file operations
     (moving, copying or deleting files or directories) in a
     background thread.

     The class works by setting up a list of tasks, then running them
     with run().

     Error handling is not terribly advanced - any error will cause
     the task to fail, and abort any operations set to follow it.

     In move and delete operations, the source and destination paths
     may refer to either single files or directories, whatever makes
     sense.

     Copying is currently only implemented for files, not directories.
   */
  enum FileOpTaskOps { FILEOP_COPY, FILEOP_MOVE, FILEOP_DELETE };

  struct FileOpTask : Jobify::Job
  {
    struct FileOp
    {
      int type;
      std::string source, dest; // Use 'dest' for delete operations
    };

    FileOpTask(Jobify::JobInfoPtr info = Jobify::JobInfoPtr())
      : Jobify::Job(info) {}

    void addOp(const FileOp &op);
    void addOp(int type, const std::string &source, const std::string &dest);

    // Convenience functions for adding operations.
    void copy(const std::string &source, const std::string &dest)
    { addOp(FILEOP_COPY, source, dest); }
    void move(const std::string &source, const std::string &dest)
    { addOp(FILEOP_MOVE, source, dest); }
    void del(const std::string &target)
    { addOp(FILEOP_DELETE, "", target); }

  private:
    void doJob();
    typedef std::list<FileOp> OpList;
    OpList ops;
  };
}

#endif
