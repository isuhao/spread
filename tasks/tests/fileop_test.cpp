#include "fileop.hpp"

#include <iostream>
using namespace std;
using namespace Tasks;
using namespace Jobify;

void testStatus(Job &j)
{
  JobInfoPtr info = j.getInfo();

  if(info->isBusy()) cout << "Busy!";
  else if(!info->hasStarted()) cout << "Not started yet!";
  else if(info->isSuccess()) cout << "Success!";
  else if(info->isNonSuccess()) cout << "Failure: " << info->message;
  cout << "  - progress " << info->getCurrent() << "/" << info->getTotal() << endl;
}

void testJob(Job &j)
{
  j.run();
  testStatus(j);
}

int main()
{
  {
    cout << "Empty test:\n";
    FileOpTask ops;
    testStatus(ops);
    testJob(ops);
  }

  {
    cout << "\nCopy:\n";
    FileOpTask ops;
    ops.copy("test.sh", "_tmp.out");
    testJob(ops);
  }

  {
    cout << "\nCopy - Remove - Move:\n";
    FileOpTask ops;
    ops.copy("_tmp.out", "_tmp2.out");
    ops.del("_tmp.out");
    ops.move("_tmp2.out", "_tmp3.out");
    testJob(ops);
  }

  {
    cout << "\nCopy to dir:\n";
    FileOpTask ops;
    ops.copy("_tmp3.out", "_outdir1/");
    testJob(ops);
  }

  return 0;
}
