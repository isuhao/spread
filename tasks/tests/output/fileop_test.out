Empty test:
Not started yet!  - progress 0/0
Success!  - progress 0/0

Copy:
Success!  - progress 1/1

Copy - Remove - Move:
Success!  - progress 2/3

Copy to dir:
Failure: Cannot copy to directories yet  - progress 0/1
